### PR TITLE
Parallel sublist

### PR DESCRIPTION
**Pull request by EvgeniiaVak**
- [x] parallelization with std-only tools
  - [x] simple: launch as many threads as there are windows
  - [x] sensible:
    - [x] no more threads than cpu at the same time
    - [x] only launch the comparison if we did not find a match yet
- [x] rayon
- [ ] ~~tokio~~ (I decided not to try tokio on this exercise, because as they say in [the tutorial](https://tokio.rs/tokio/tutorial): 
    > Tokio is designed for IO-bound applications where each individual task spends most of its time waiting for IO. If the only thing your application does is run computations in parallel, you should be using [rayon](https://docs.rs/rayon/). 
- [x] perf tests

### CODE REVIEW

File: sublist/src/lib.rs [hunk 1]
```diff
@@ -6,7 +12,54 @@ pub enum Comparison {
...
+            thread_counter += 1;
+            // create a new transmitter for each thread
+            let tx = tx.clone();
+
+            let b = b.clone();
+
+            s.spawn(move || {
+                // TODO: kill the thread if we already found a match
+                let are_equal = *b == window;
+                tx.send(are_equal);
```
Comment (senekor) [inline sublist/src/lib.rs]: Clippy is warning that this `Result` must be used. Do you have clippy setup in your editor? I find it extremely helpful.

**Review by senekor — COMMENTED** [review]
I think the comments you left are quite correct: Before the threads based solution can be more performant than the serial one, I expect some thread pooling will be necessary. Creating one thread per window is a very high overhead.

File: sublist/src/lib.rs (hunk 1 above)
Comment (EvgeniiaVak) [reply sublist/src/lib.rs]: Yes, I see that warning... I don't know how to resolve it though. I tried to google why that's happening, even tried to check every result (errors including) below like this: 

```rust
        for _ in 0..thread_counter {
            let res = rx.recv();
            match res {
                Ok(are_equal) => {
                    if are_equal {
                        return true;
                    }
                }
                Err(_) => {}
            }
        }
```

But the warning persist. Do you know how it supposed to be handled?

Comment (senekor) [reply sublist/src/lib.rs]: I think you're looking at the wrong lines of code, it's where you spawn the threads, not where you read from the channel:
```rust
s.spawn(move || {
    // TODO: kill the thread if we already found a match
    let are_equal = *b == window;
    tx.send(are_equal).unwrap();
});
```
A quick-and-dirty `.unwrap()` will do the trick.

Comment (EvgeniiaVak) [reply sublist/src/lib.rs]: Aaah, no I was looking at this line all right, but I thought I should handle it somewhere in the receiver end :D ok, thanks, fixing.

File: sublist/src/lib.rs [hunk 2]
```diff
@@ -6,22 +12,181 @@ pub enum Comparison {
...
+    /// # Panics
+    ///
+    /// The `new` function will panic if the size is zero.
+    pub fn new(size: usize, scope: &'s thread::Scope<'s, '_>) -> ScopedThreadPool<'s> {
+        assert!(size > 0);
+
+        let mut workers = Vec::with_capacity(size);
+        let (sender, receiver) = mpsc::channel();
+
+        let receiver = Arc::new(Mutex::new(receiver));
```
Comment (senekor) [inline sublist/src/lib.rs]: This is unfortunate, it means that your threads have to synchronize on getting new input. Your use case really calls for a single-producer-multi-consumer channel, which the standard library does not provide.

There are several channel crates out there which do provide that functionality, I know of crossbeam, flume and kanal. [crossbeam](https://docs.rs/crossbeam/latest/crossbeam/channel/index.html) is probably the most mature and "standard" one, so I recommend going with that for a start. But it's good to know there are others.

**Comment by senekor** [comment]
Thread pooling is a difficult topic, I gotta spend some more time on this to make a good review...

*robotic corporate voice*: Thank you for your patience :innocent:

File: sublist/src/lib.rs (hunk 2 above)
Comment (senekor) [reply sublist/src/lib.rs]: Now that I think about it, you wanted to use only the standard library for this part, right? In that case, maybe consider not using channels at all. One approach that would probably work is to wrap the iterator over windows in an `Arc<Mutex<_>>`. Haven't given it too much thought though.

Comment (EvgeniiaVak) [reply sublist/src/lib.rs]: Yeah... I tried to abandon the threadpool idea at all and create `Arc<Mutex<VecDeque<Job>>>` in the last commit, but it did not work out also. Well, rayon somehow did it, and it has a lot of lifetime annotation in its classes, so maybe the last version (after [612fe9e](https://github.com/EvgeniiaVak/Rust-Exercism/pull/1/commits/612fe9e88dee21413c986bbcf2dfd9f89bc89052)) will be working if putting some annotations in the right places 🤔 so far the ones I've tried did not work though (I did not even push any).

**Comment by EvgeniiaVak** [comment]
Because I'm quite stuck with this std-only tools idea, I'll leave this PR dandling for now and will return to it later, after I have some more experience with Rust lifetimes and threads. Any improvements are welcome 🙂 
Thank you, @remlse 🙇

**Comment by EvgeniiaVak** [comment]
Ok, I managed to get it working by returning thread::scope and removing the Box around task functions.

File: sublist/src/lib.rs (hunk 2 above)
Comment (EvgeniiaVak) [reply sublist/src/lib.rs]: Ok, I managed to get it working by returning thread::scope and removing the Box around task functions.

**Comment by EvgeniiaVak** [comment]
Well, I'm obviously doing something wrong with the threads version 😆 
test_result

**Comment by senekor** [comment]
Hahaha, well, I consider this a success!! That's exactly why we do benchmarking, right? :wink:

Great job on getting the borrow checker to play nice :muscle:

**Comment by EvgeniiaVak** [comment]
I couldn't get profiling working with criterion and pprof on my mac, but I stumbled upon this video https://youtu.be/JRMOIE_wAFk?t=1860 about profiling in clion IDE and it's really easy to do that there, so I created tests specifically to run profiling on them and then used used Clion for profiling. Unfortunately I couldn't find how to save the results as interactive svg, but at least it could save in png format.

**Comment by EvgeniiaVak** [comment]
@remlse 
I fixed it 🎉  
The problem was I created a task queue sequentially 😆 which defeated the whole point and was not even needed. Ok, lesson learned.

violin

**Review by senekor — APPROVED** [review]
Fantastic!! And it looks like the performance is even a little better than rayon. Great success :smiley:

**Pull request merged**